iroh/examples/transfer.rs

```diff
@@ -0,0 +1,313 @@
+            }
+            Ok::<_, anyhow::Error>(())
+        });
+    }
+
+    // stop with SIGINT (ctrl-c)
+    Ok(())
+}
+
+async fn fetch(ticket: &str, relay_url: Option<String>) -> anyhow::Result<()> {
+    let ticket: NodeTicket = ticket.parse()?;
+    let secret_key = SecretKey::generate();
+    let relay_mode = match relay_url {
+        Some(relay_url) => {
+            let relay_url = RelayUrl::from_str(&relay_url)?;
+            let relay_map = RelayMap::from_url(relay_url);
+            RelayMode::Custom(relay_map)
+        }
+        None => RelayMode::Default,
+    };
+    let endpoint = Endpoint::builder()
+        .secret_key(secret_key)
+        .alpns(vec![TRANSFER_ALPN.to_vec()])
+        .relay_mode(relay_mode)
+        .bind()
+        .await?;
+
+    let start = Instant::now();
+
+    let me = endpoint.node_id();
+    println!("node id: {me}");
+    println!("node listening addresses:");
+    for local_endpoint in endpoint
+        .direct_addresses()
+        .next()
+        .await
+        .context("no endpoints")?
+    {
+        println!("\t{}", local_endpoint.addr)
+    }
+
+    let relay_url = endpoint
+        .home_relay()
+        .expect("should be connected to a relay server, try calling `endpoint.local_endpoints()` or `endpoint.connect()` first, to ensure the endpoint has actually attempted a connection before checking for the connected relay server");
+    println!("node relay server url: {relay_url}\n");
+
+    // Attempt to connect, over the given ALPN.
+    // Returns a Quinn connection.
+    let conn = endpoint
+        .connect(ticket.node_addr().clone(), TRANSFER_ALPN)
+        .await?;
+    info!("connected");
+
+    // Use the Quinn API to send and recv content.
+    let (mut send, mut recv) = conn.open_bi().await?;
+
+    let message = format!("{me} is saying 'hello!'");
+    send.write_all(message.as_bytes()).await?;
+
+    // Call `finish` to close the send side of the connection gracefully.
+    send.finish()?;
+
+    let (len, ttfb, chnk) = drain_stream(&mut recv, false).await?;
+
+    // We received the last message: close all connections and allow for the close
+    // message to be sent.
+    endpoint.close(0u8.into(), b"bye").await?;
+
+    let duration = start.elapsed();
+    println!(
+        "Received {} B in {:.4}s with ttfb {}s in {} chunks",
+        HumanBytes(len as u64),
+        duration.as_secs_f64(),
+        ttfb.as_secs_f64(),
+        chnk
+    );
+    println!(
+        "Transferred {} in {:.4}, {}/s",
+        HumanBytes(len as u64),
+        duration.as_secs_f64(),
+        HumanBytes((len as f64 / duration.as_secs_f64()) as u64)
+    );
+
+    Ok(())
+}
+
+async fn drain_stream(
+    stream: &mut iroh_net::endpoint::RecvStream,
+    read_unordered: bool,
+) -> Result<(usize, Duration, u64)> {
+    let mut read = 0;
+
+    let download_start = Instant::now();
+    let mut first_byte = true;
+    let mut ttfb = download_start.elapsed();
```

Comment:
FWIW I have no idea what `ttfb` stands for...

Reply:
time to first byte probably

Reply:
time to first byte :). I'll make sure the vars are more readable. TY